### ISSUE TITLE
Macros apparently have higher precedence than numbers

### ISSUE THREAD

**Issue by Y-Less** [Zeex/pawn#183]
```pawn
#define xFF
new a = 0xFF;
printf("%d", a); // Prints 0
```

That probably shouldn't happen, but only if the symbol is a valid part of a number and used as such.  I quite like this, that I saw in a recent include:

```pawn
#define _Seconds * 1000
new a = 5_Seconds;
printf("%d", a); // Prints 5000
```

In that case, `5_` is not possibly a valid number so should be replaced; while things like `0x`, `0b`, or `5e` could be so shouldn't be.

**Comment by Y-Less** [comment]
http://forum.sa-mp.com/showthread.php?p=3924669#post3924669
http://forum.sa-mp.com/showthread.php?t=640660

**Comment by VVWVV** [comment]
The preprocessor works first, and it does not analize the code. To fix the problem we need to rewrite the preprocessor.

**Comment by Y-Less** [comment]
> I guess this is not a bug. 

> To fix the problem

Which is it?  If I write:

```pawn
#define xFF
new AxFF = 255;
new a = AxFF;
printf("%d", a); // Prints 255
```

That is fine, because macros don't match in the middle of symbols, so they probably shouldn't match in the middle of valid numbers either.

**Comment by VVWVV** [comment]
Sorry) Yeah, the preprocessor skips all non-alphabetic characters here: https://github.com/Zeex/pawn/blob/master/source/compiler/sc2.c#L1772-L1780

**Comment by Dayvison** [comment]
Don't skip if have `a-z @ _` before the number

```pawn
#define test -

0test
10test
_0test
```
Output:
```
0-
10-
_0test
```

**Comment by Y-Less** [comment]
`_` and `@` should be counted - they can start legitimate tokens.  I literally have a macros called `___1` and `@1`.

**Comment by Dayvison** [comment]
Yeah is just a example, what i want to say is if as any character before number, don't match.

**Comment by Dayvison** [comment]
What i tried
```pawn
#define hook%1(%2) forward UNIQUE_FUNCTION<@yH0%1@...>(%2);UNIQUE_FUNCTION<@yH0%1@...>(%2)
// Strip out extra spaces (nicely recursive)
#define @yH0%0\32;%1(%2) @yH0%1(%2)

#define tag:: tag_
hook tag::function()
